File: ambeth/Ambeth.Bytecode/ambeth/bytecode/behavior/IBytecodeBehaviorState.cs
using De.Osthus.Ambeth.Ioc;
using De.Osthus.Ambeth.Typeinfo;
using De.Osthus.Ambeth.Util;
using System;
using System.Reflection;

namespace De.Osthus.Ambeth.Bytecode.Behavior
{
    public interface IBytecodeBehaviorState
    {
	    Type OriginalType { get; }

	    Type CurrentType { get; }

	    NewType NewType { get; }

        IServiceContext BeanContext { get; }

	    IEnhancementHint Context { get; }
        
	    T GetContext<T>() where T : IEnhancementHint;

        Object GetContext(Type contextType);

        PropertyInstance GetProperty(String propertyName);

	    MethodInstance[] GetAlreadyImplementedMethodsOnNewType();

	    FieldInstance GetAlreadyImplementedField(String fieldName);

        bool HasMethod(MethodInstance method);

	    bool IsMethodAlreadyImplementedOnNewType(MethodInstance method);
    }
}

File: jambeth/jambeth-bytecode/src/main/java/de/osthus/ambeth/bytecode/behavior/IBytecodeBehaviorState.java
package de.osthus.ambeth.bytecode.behavior;

import de.osthus.ambeth.bytecode.FieldInstance;
import de.osthus.ambeth.bytecode.IEnhancementHint;
import de.osthus.ambeth.bytecode.MethodInstance;
import de.osthus.ambeth.bytecode.PropertyInstance;
import de.osthus.ambeth.ioc.IServiceContext;
import de.osthus.ambeth.repackaged.org.objectweb.asm.Type;

public interface IBytecodeBehaviorState
{
	Class<?> getOriginalType();

	Class<?> getCurrentType();

	Type getNewType();

	IServiceContext getBeanContext();

	IEnhancementHint getContext();

	<T extends IEnhancementHint> T getContext(Class<T> contextType);

	PropertyInstance getProperty(String propertyName);

	MethodInstance[] getAlreadyImplementedMethodsOnNewType();

	FieldInstance getAlreadyImplementedField(String fieldName);

	boolean hasMethod(MethodInstance method);

	boolean isMethodAlreadyImplementedOnNewType(MethodInstance method);
}
